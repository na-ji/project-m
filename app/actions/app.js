export const GOING_BACK = 'GOING_BACK';

export function goingBack () {
    return {
        type: GOING_BACK
    };
}
