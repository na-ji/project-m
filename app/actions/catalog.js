export const GET_POPULAR_MANGAS = 'GET_POPULAR_MANGAS';
export const LOAD_MORE = 'LOAD_MORE';
export const SEARCH = 'SEARCH';
export const RESET_SEARCH = 'RESET_SEARCH';
export const RECEIVE_MANGAS_LIST = 'RECEIVE_MANGAS_LIST';
export const RECEIVE_MANGA_DETAILS = 'RECEIVE_MANGA_DETAILS';
// export const CANCEL_REQUEST = 'CANCEL_REQUEST';
// export const REQUEST_CANCELED = 'REQUEST_CANCELED';

import MangaManager from '../utils/manga-manager';
import CatalogManager from '../utils/catalog-manager';
import _ from 'lodash';

function receiveMangasList (response, override) {
    return {
        type: RECEIVE_MANGAS_LIST,
        response,
        override
    };
}

function receiveMangaDetails (manga) {
    return {
        type: RECEIVE_MANGA_DETAILS,
        manga
    };
}

function getPopularMangas (catalogName) {
    let catalog = CatalogManager.getCatalog(catalogName);

    return {
        type: GET_POPULAR_MANGAS,
        catalogName,
        catalog
    };
}

function loadMore () {
    return {
        type: LOAD_MORE
    };
}

function search (query) {
    return {
        query,
        type: SEARCH
    };
}

export function resetSearch () {
    return {
        type: RESET_SEARCH
    };
}

function fetchMangasList (promise, override) {
    return (dispatch) => {
        promise.then(function (response) {
            dispatch(receiveMangasList(response, override));

            _.forEach(response.promises, function (promise) {
                promise.then(function (manga) {
                    dispatch(receiveMangaDetails(manga));
                });
            });
        });
    };
}

export function fetchPopularMangas (catalogName) {
    return (dispatch, getState) => {
        const oldCatalogName = getState().catalog.catalogName;
        if (catalogName !== oldCatalogName) {
            dispatch(getPopularMangas(catalogName));
            dispatch(fetchMangasList(MangaManager.getPopularManga(catalogName), true));
        }
    };
}

export function fetchMore () {
    return (dispatch, getState) => {
        const state = getState().catalog;
        if (state.hasNext) {
            dispatch(loadMore());
            dispatch(fetchMangasList(MangaManager.getPopularManga(state.catalogName, true), false));
        }
    };
}

export function searchManga (query) {
    return (dispatch, getState) => {
        const oldQuery = getState().catalog.query;
        if (query !== oldQuery) {
            dispatch(search(query));
            dispatch(fetchMangasList(MangaManager.searchManga(getState().catalog.catalogName, query)));
        }
    };
}
