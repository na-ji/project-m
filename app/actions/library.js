import Promise from 'bluebird';
import { ipcRenderer } from 'electron';

import MangaManager from '../utils/manga-manager';

export const REQUEST_LIBRARY = 'REQUEST_LIBRARY';
export const RECEIVE_LIBRARY = 'RECEIVE_LIBRARY';
export const REFRESH_LIBRARY = 'REFRESH_LIBRARY';
export const REFRESH_PROGRESS = 'REFRESH_PROGRESS';
export const TOGGLE_MANGA_TO_LIBRARY = 'TOGGLE_MANGA_TO_LIBRARY';

function requestLibrary () {
    return {
        type: REQUEST_LIBRARY
    };
}

function receiveLibrary (mangas) {
    return {
        type: RECEIVE_LIBRARY,
        mangas
    };
}

export function toggleMangaToLibrary (manga) {
    return (dispatch) => {
        dispatch({
            type: TOGGLE_MANGA_TO_LIBRARY,
            manga
        });

        MangaManager.toggleInLibrary(manga);
    };
}

export function fetchLibrary () {
    return (dispatch, getState) => {
        const state = getState().library;

        if (!state.loaded) {
            dispatch(requestLibrary());
            MangaManager.getLibrary().observe().subscribe(mangas => {
                dispatch(receiveLibrary(mangas));
            });
        }
    };
}

export function refreshLibrary () {
    return (dispatch, getState) => {
        const state = getState().library;

        if (state.loaded && state.mangas.length > 0) {
            dispatch({
                type: REFRESH_LIBRARY
            });

            Promise.mapSeries(state.mangas, (manga) => {
                const promise = MangaManager.getChapterList(manga);

                promise.then(() => {
                    const state = getState().library;

                    ipcRenderer.send('progress-bar', state.progressPercent / 100);

                    dispatch({
                        type: REFRESH_PROGRESS
                    });
                });

                return promise;
            }).then(values => {
                new Notification('Library refreshed', {// eslint-disable-line
                    body: 'X new chapters found'
                });

                ipcRenderer.send('progress-bar', -1);
            });
        }
    };
}
