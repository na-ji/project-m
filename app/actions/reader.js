export const REQUEST_PAGES_URL = 'REQUEST_PAGES_URL';
export const RECEIVE_PAGES_URL = 'RECEIVE_PAGES_URL';
export const CHANGE_PAGE = 'CHANGE_PAGE';
export const REQUEST_IMAGE_URL = 'REQUEST_IMAGE_URL';
export const RECEIVE_IMAGE_URL = 'RECEIVE_IMAGE_URL';
export const RECEIVE_IMAGE = 'RECEIVE_IMAGE';
export const REQUEST_IMAGE = 'REQUEST_IMAGE';
export const INIT_IMAGES = 'INIT_IMAGES';
export const CANCEL_REQUEST = 'CANCEL_REQUEST';
export const REQUEST_CANCELED = 'REQUEST_CANCELED';

import mangaManager from '../core/manga-manager';
import _ from 'lodash';

function requestPagesUrl () {
    return {
        type: REQUEST_PAGES_URL
    };
}

function receivePagesUrl (pagesUrl, chapterId) {
    return {
        type: RECEIVE_PAGES_URL,
        pagesUrl,
        chapterId
    };
}

export function changePage (nextPage) {
    return {
        type: CHANGE_PAGE,
        nextPage
    };
}

function requestImageUrl (pageUrl) {
    return {
        type: REQUEST_IMAGE_URL,
        pageUrl
    };
}

function receiveImageUrl (pageUrl, imageUrl) {
    return {
        type: RECEIVE_IMAGE_URL,
        pageUrl,
        imageUrl
    };
}

function requestImage (image) {
    return {
        type: REQUEST_IMAGE,
        image
    };
}

function receiveImage (image, htmlImage) {
    return {
        type: RECEIVE_IMAGE,
        image,
        htmlImage
    };
}

function initImages (images) {
    return {
        type: INIT_IMAGES,
        images
    };
}

export function cancelRequest () {
    return {
        type: CANCEL_REQUEST
    };
}

function requestCanceled () {
    return {
        type: REQUEST_CANCELED
    };
}

function fetchImage (manga) {
    return (dispatch, getState) => {
        const { pages, images } = getState().reader;

        if (_.findIndex(images.images, _.isUndefined) === -1) {
            // if we already loaded all the images, we stop here
            return;
        }

        let downloadingImage = new Image();
        downloadingImage.onload = function () {
            // When the image has been downloaded by the browser
            const { images } = getState().reader;
            dispatch(receiveImage(images.imageFetching, downloadingImage));

            if (!images.cancelRequest && _.findIndex(images.images, _.isUndefined) > -1) {
                // if there is still image to load
                dispatch(fetchImage(manga));
            } else if (images.cancelRequest) {
                dispatch(requestCanceled());
            }
        };

        let pageUrl = pages.pagesUrl[images.imageFetching];
        dispatch(requestImageUrl(pageUrl));

        mangaManager.getImageURL(manga, pageUrl).then(function (imageURL) {
            dispatch(receiveImageUrl(pageUrl, imageURL));

            const { images } = getState().reader;
            if (!images.cancelRequest) {
                dispatch(requestImage(imageURL));

                // we start the downloading of the image by adding it to the Image object
                downloadingImage.src = imageURL;
            } else {
                dispatch(requestCanceled());
            }
        }).catch(function (error) {
            console.log(error);
            // Todo: manage timeouts
        });
    };
}

function fetchPages (manga, chapter) {
    return (dispatch, getState) => {
        dispatch(requestPagesUrl());
        dispatch(initImages([]));

        mangaManager.getChapterPages(manga, chapter).then(function (pagesUrl) {
            dispatch(receivePagesUrl(pagesUrl, chapter.id));
            const { pages } = getState().reader;

            if (pagesUrl.length) {
                if (pages.cancelRequest) {
                    dispatch(requestCanceled());
                } else {
                    dispatch(initImages(new Array(pagesUrl.length)));
                    dispatch(fetchImage(manga));
                }
            }
        }).catch(function (error) {
            console.log(error);
            // Todo: manage timeouts
        });
    };
}

export function fetchPagesIfNeeded (manga, chapter) {
    return (dispatch, getState) => {
        const { pages } = getState().reader;

        // If the request was canceled previously and requestCanceled never has been dispatched
        if (pages.cancelRequest) {
            dispatch(requestCanceled());
        }

        if (pages.chapterId !== chapter.id) {
            // If it is a new chapter
            dispatch(fetchPages(manga, chapter));
        } else {
            // If we already loaded the chapter
            dispatch(fetchImage(manga));
        }
    };
}
