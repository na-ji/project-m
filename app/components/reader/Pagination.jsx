'use babel';
import React from 'react';

class Pagination extends React.Component {
    render () {
        let self = this;
        let pagesDisplayed = [];
        let numberDisplayed = 20;
        let offset = 0;
        // calculate the position of the '...'
        if (this.props.pages.length > numberDisplayed) {
            if (this.props.page > numberDisplayed / 2) {
                pagesDisplayed.push('...');
            }

            offset = Math.max(this.props.page - numberDisplayed / 2, 0);
            pagesDisplayed = pagesDisplayed.concat(
                this.props.pages.slice(
                    offset,
                    Math.min(offset + numberDisplayed, this.props.pages.length - 1)
                )
            );

            if (this.props.page + numberDisplayed / 2 + 1 < this.props.pages.length) {
                pagesDisplayed.push('...');
            }
        } else {
            pagesDisplayed = this.props.pages;
        }

        return (
            <div>
                <nav>
                    <ul className="pagination">
                        <li className={(self.props.page === 0 ? 'disabled' : '')} onClick={this.props.handler} data-page="previous">
                            <a href="#!">
                                <span>&laquo;</span>
                            </a>
                        </li>
                        {pagesDisplayed.map(function (page, index) {
                            if (page === '...') {
                                return (
                                    <li key={index} className="disabled" data-page="none" onClick={self.props.handler}>
                                        <a href="#!">...</a>
                                    </li>
                                );
                            }
                            return (
                                <li key={index} className={(self.props.page === offset + index ? 'active ' : '') + (self.props.loadedImages[offset + index] ? 'loaded' : '')} onClick={self.props.handler} data-page={offset + index}>
                                    <a href="#!">{offset + index + 1}</a>
                                </li>
                            );
                        })}
                        <li className={(self.props.page === this.props.pages.length - 1 ? 'disabled' : '')} onClick={this.props.handler} data-page="next">
                            <a href="#!">
                                <span>&raquo;</span>
                            </a>
                        </li>
                    </ul>
                </nav>
            </div>
        );
    }
}

module.exports = Pagination;
