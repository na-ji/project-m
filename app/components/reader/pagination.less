.pagination > .loaded {
    > a {
        border-color: #4caf50;
    }

    + li > a {
        border-left-color: #4caf50;
    }
}

.imgPage {
    display: inline;

    &:hover {
        cursor: pointer;
    }
}
