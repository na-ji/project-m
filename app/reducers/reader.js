import { combineReducers } from 'redux';
import {
    REQUEST_PAGES_URL, RECEIVE_PAGES_URL, CHANGE_PAGE,
    REQUEST_IMAGE_URL, RECEIVE_IMAGE_URL,
    REQUEST_IMAGE, RECEIVE_IMAGE, INIT_IMAGES, CANCEL_REQUEST, REQUEST_CANCELED
} from '../actions/reader';
import _ from 'lodash';

function pages (state = {
    currentPage: 0,
    isFetching: false,
    chapterId: '',
    pagesUrl: [],
    promise: null
}, action) {
    switch (action.type) {
        case REQUEST_PAGES_URL:
            return Object.assign({}, state, {
                isFetching: true,
                pagesUrl: [],
                currentPage: 0,
                promise: action.promise
            });
        case CANCEL_REQUEST:
            if (state.promise) {
                state.promise.cancel();
            }

            return Object.assign({}, state, {
                promise: null
            });
        case RECEIVE_PAGES_URL:
            return Object.assign({}, state, {
                isFetching: false,
                pagesUrl: action.pagesUrl,
                chapterId: action.chapterId,
                promise: null
            });
        case CHANGE_PAGE:
            return Object.assign({}, state, {
                currentPage: action.nextPage
            });
        default:
            return state;
    }
}

function images (state = {
    imageFetching: 0,
    isFetching: false,
    images: [],
    cancelRequest: false,
    promise: null
}, action) {
    switch (action.type) {
        case REQUEST_IMAGE_URL:
            return Object.assign({}, state, {
                isFetching: true,
                promise: action.promise
            });
        case RECEIVE_IMAGE_URL:
            return Object.assign({}, state, {
                isFetching: false,
                promise: null
            });
        case REQUEST_IMAGE:
            return Object.assign({}, state, {
                isFetching: true
            });
        case INIT_IMAGES:
            return Object.assign({}, state, {
                images: action.images,
                imageFetching: 0,
                cancelRequest: false
            });
        case CANCEL_REQUEST:
            let nextState = {
                promise: null
            };

            if (state.promise) {
                state.promise.cancel();
            } else if (state.isFetching) {
                nextState.cancelRequest = true;
            }

            return Object.assign({}, state, nextState);
        case REQUEST_CANCELED:
            return Object.assign({}, state, {
                cancelRequest: false
            });
        case CHANGE_PAGE:
            // If the image is not yet loaded, we set it as the next image to load
            if (!state.images[action.nextPage]) {
                return Object.assign({}, state, {
                    nextImageFetching: action.nextPage
                });
            }
            return state;
        case RECEIVE_IMAGE:
            let nextImageFetching = state.imageFetching;

            if (!_.isUndefined(state.nextImageFetching)) {
                nextImageFetching = state.nextImageFetching;
            } else if (_.findIndex(state.images, _.isUndefined) > -1) {
                // We select the next image not downloaded
                if (_.findIndex(state.images, _.isUndefined, state.imageFetching + 1) > -1) {
                    // if there is an image not downloaded AFTER the current image
                    nextImageFetching = _.findIndex(state.images, _.isUndefined, state.imageFetching + 1);
                } else {
                    // if there is an image not downloaded BEFORE the current image
                    nextImageFetching = _.findIndex(state.images, _.isUndefined);
                }
            }

            let newState = Object.assign({}, state, {
                isFetching: false,
                imageFetching: nextImageFetching,
                images: [
                    ...state.images.slice(0, state.imageFetching),
                    action.htmlImage,
                    ...state.images.slice(state.imageFetching + 1)
                ]
            });

            if (state.nextImageFetching !== undefined) {
                // We remove the nextImageFetching once the loading started
                newState = Object.assign({}, newState, {
                    nextImageFetching: undefined
                });
            }

            return newState;
        default:
            return state;
    }
}

const rootReducer = combineReducers({
    pages,
    images
});

export default rootReducer;
