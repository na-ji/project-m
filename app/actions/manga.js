import _ from 'lodash';

import MangaManager from '../utils/manga-manager';
import { TOGGLE_MANGA_TO_LIBRARY } from './library';

export const LOAD_MANGA = 'LOAD_MANGA';
// export const REQUEST_CHAPTERS = 'REQUEST_CHAPTERS';
export const RECEIVE_CHAPTERS = 'RECEIVE_CHAPTERS';
// export const REQUEST_DETAILS = 'REQUEST_DETAILS';
export const RECEIVE_DETAILS = 'RECEIVE_DETAILS';
export const MANGA_TOGGLE_LIBRARY = 'MANGA_TOGGLE_LIBRARY';

function loadManga (manga) {
    return {
        type: LOAD_MANGA,
        manga
    };
}

function receiveChapters (chapters) {
    return {
        type: RECEIVE_CHAPTERS,
        chapters
    };
}

function receiveDetails (manga) {
    return {
        type: RECEIVE_DETAILS,
        manga
    };
}

export function toggleLibrary () {
    return (dispatch, getState) => {
        dispatch({
            type: MANGA_TOGGLE_LIBRARY
        });

        const { manga } = getState().manga;
        MangaManager.persistManga(manga);

        dispatch({
            type: TOGGLE_MANGA_TO_LIBRARY,
            manga
        });
    };
}

function fetchInfos (manga) {
    return (dispatch) => {
        dispatch(loadManga(manga));

        if (_.isNil(manga.chapters) || !manga.chapters.length) {
            MangaManager.getChapterList(manga).then(function (chapters) {
                dispatch(receiveChapters(chapters));
            });
        } else {
            MangaManager.getMangaById(manga.id).then(function (response) {
                dispatch(receiveChapters(response.chapters));
            });
        }

        if (!manga.detailsFetched) {
            MangaManager.getMangaDetail(manga).then(function (manga) {
                dispatch(receiveDetails(manga));
            });
        }
    };
}

export function fetchInfosIfNeeded (manga) {
    return (dispatch, getState) => {
        const oldManga = getState().manga.manga;

        if (oldManga.id !== manga.id) {
            dispatch(fetchInfos(manga));
        }
    };
}
