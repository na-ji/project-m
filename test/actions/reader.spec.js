import { spy } from 'sinon';

jest.mock('electron-rxdb');
jest.mock('../../app/utils/manga-manager');

import * as actions from '../../app/actions/reader';

describe('reader actions', () => {
    it('should cancel request', () => {
        expect(actions.cancelRequest()).toMatchSnapshot();
    });

    it('should fetch pages if needed', (done) => {
        const manga = {id: 'manga'};
        const chapter = {id: 'chapter'};
        const fetchPagesIfNeeded = actions.fetchPagesIfNeeded(manga, chapter);

        const dispatch = spy();
        const getState = () => ({ reader: { pages: { chapterId: null } } });
        fetchPagesIfNeeded(dispatch, getState);

        const fetchPages = dispatch.args[0][0];
        const dispatch2 = spy();
        fetchPages(dispatch2);
        setTimeout(() => {
            expect(dispatch2.callCount).toBe(5);
            expect(dispatch2.calledWithMatch({ type: actions.REQUEST_PAGES_URL })).toBe(true);
            expect(dispatch2.calledWithMatch({ type: actions.INIT_IMAGES, images: [] })).toBe(true);
            expect(dispatch2.calledWithMatch({
                type: actions.RECEIVE_PAGES_URL,
                pagesUrl: new Array(5),
                chapterId: 'chapter'
            })).toBe(true);
            expect(dispatch2.calledWithMatch({
                type: actions.INIT_IMAGES,
                images: new Array(5)
            })).toBe(true);
            done();
        }, 100);
    });

    // it('should fetch images if pages are already fetched', (done) => {
    //     const manga = {id: 'manga'};
    //     const chapter = {id: 'chapter'};
    //     const fetchPagesIfNeeded = actions.fetchPagesIfNeeded(manga, chapter);
    //
    //     const dispatch = spy();
    //     const getState = () => ({ reader: { pages: { chapterId: 'chapter' } } });
    //     fetchPagesIfNeeded(dispatch, getState);
    //
    //     const fetchImage = dispatch.args[0][0];
    //     const dispatch2 = spy();
    //     const getState2 = () => ({
    //         reader: {
    //             pages: { chapterId: 'chapter', pagesUrl: ['https://swag.army/1', 'https://swag.army/2'] },
    //             images: { images: [], cancelRequest: false, imageFetching: 0 }
    //         }
    //     });
    //     fetchImage(dispatch2, getState2);
    //     setTimeout(() => {
    //         console.log(dispatch2.args);
    //         console.log(dispatch2.callCount);
    //         done();
    //     }, 100);
    // });
});
