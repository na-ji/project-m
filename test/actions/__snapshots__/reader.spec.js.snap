// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reader actions should cancel request 1`] = `
Object {
  "type": "CANCEL_REQUEST",
}
`;
